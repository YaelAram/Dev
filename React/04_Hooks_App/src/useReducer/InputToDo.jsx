import React, { useRef } from 'react';

export const InputToDo = React.memo(
    ( { handleNewToDo } ) => {
        const titleInput = useRef();

        const handleSubmit = ( evt ) => {
            evt.preventDefault();
            const title = titleInput.current.value.trim();
            if( title ){
                titleInput.current.value = '';
                handleNewToDo( title );
            }
        };

        return (
            <form onSubmit={ handleSubmit } aria-label="form">
                <input type="text" placeholder="Buy Milk" ref={ titleInput } autoFocus />
                <input type="submit" value="Create"/>
            </form>
        );
    }
);
