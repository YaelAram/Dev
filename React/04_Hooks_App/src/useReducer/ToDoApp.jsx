import React from 'react';
import { ToDoList, InputToDo } from './';
import { useToDo } from '../hooks';

export function ToDoApp() {
    const { todos, handleNewToDo, handleRemoveToDo, handleToggleToDo } = useToDo();

    return (
        <section>
            <h1>ToDo App</h1>
            <InputToDo handleNewToDo={ handleNewToDo }/>
            <ToDoList todos={ todos } handleRemoveToDo={ handleRemoveToDo } handleToggleToDo={ handleToggleToDo }/>
        </section>
    );
};
