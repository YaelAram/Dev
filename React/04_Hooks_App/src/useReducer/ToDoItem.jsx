import React from 'react';

export const ToDoItem = React.memo(
    ( { id, title, done, handleRemoveToDo, handleToggleToDo } ) => {
        const handleClickRemove = () => handleRemoveToDo( id );
        const handleCheckBoxToggle = () => handleToggleToDo( id );

        return (
            <li>
                <input type="checkbox" checked={ done } onChange={ handleCheckBoxToggle }/>
                <span className={ ( done ) ? "doneToDo" : "" } data-testid="title-to-do">{ title }</span>
                <input type="button" value="Remove" onClick={ handleClickRemove }/>
            </li>
        );
    }
);
