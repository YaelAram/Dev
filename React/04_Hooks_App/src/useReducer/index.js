export * from './InputToDo';
export * from './ToDoApp';
export * from './ToDoItem';
export * from './ToDoList';
export * from './toDoReducer';
