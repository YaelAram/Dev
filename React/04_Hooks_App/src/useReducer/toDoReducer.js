export const toDoReducer = ( todos, { type, payload } ) => {
    switch( type ) {
        case '[TODO] Add New ToDo':
            return [ payload, ...todos ];
        case '[TODO] Remove ToDo':
            return todos.filter( ( { id } ) => id !== payload );
        case '[TODO] Toggle ToDo':
            return todos.map( ( todo ) => {
                if( todo.id === payload ) return { ...todo, done: !todo.done };
                return todo;
            } );
        default:
            return todos;
    }
};