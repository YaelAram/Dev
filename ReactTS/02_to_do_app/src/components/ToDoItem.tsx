import { useContext } from "react";
import { ToDo } from "../interfaces/ToDo"
import { ToDoContext } from "../context/ToDoContext";

interface ToDoItemProps {
  toDo: ToDo
};

export function ToDoItem({ toDo }: ToDoItemProps) {
  const { toggleToDo } = useContext(ToDoContext);
  const { id, desc, completed } = toDo;

  const handleToggle = () => {
    toggleToDo(id);
  };

  return (
    <li onDoubleClick={handleToggle} style={{
      cursor: 'pointer',
      textDecoration: (completed) ? 'line-through' : ''
    }}>
      {desc}
    </li>
  )
};
