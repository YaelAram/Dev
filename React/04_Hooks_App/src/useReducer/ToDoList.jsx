import React from 'react';
import { ToDoItem } from './ToDoItem';

export const ToDoList = React.memo(
    ( { todos = [], handleRemoveToDo, handleToggleToDo } ) => {
        return (
            <ul>
                { 
                    todos.map( 
                        ( todo ) => 
                            <ToDoItem 
                                key={ todo.id }
                                { ...todo }
                                handleRemoveToDo={ handleRemoveToDo }
                                handleToggleToDo={ handleToggleToDo }
                            /> 
                    ) 
                }
            </ul>
        );
    }
);
