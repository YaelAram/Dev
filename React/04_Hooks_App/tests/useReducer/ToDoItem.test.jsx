import { fireEvent, render, screen } from '@testing-library/react';
import { ToDoItem } from "../../src/useReducer/ToDoItem";

describe( 'ToDoItem component', () => {
    const handleRemoveToDoMock = jest.fn();
    const handleToggleToDoMock = jest.fn();
    const toDoItem = {
        id: 1,
        title: 'Hello World',
        done: false
    };

    beforeEach( () => jest.clearAllMocks() );

    test( 'Match snapshot', () => {
        const { container } = render( 
            <ToDoItem 
                { ...toDoItem }
                handleRemoveToDo={ handleRemoveToDoMock }
                handleToggleToDo={ handleToggleToDoMock }
            /> 
        );
        expect( container ).toMatchSnapshot();
    } );

    test( 'Checkbox not selected and css class doneToDo not included', () => {
        render( 
            <ToDoItem 
                { ...toDoItem }
                handleRemoveToDo={ handleRemoveToDoMock }
                handleToggleToDo={ handleToggleToDoMock }
            /> 
        );

        expect( screen.getByTestId( 'title-to-do' ).className ).toBe( '' );
        expect( screen.getByRole( 'checkbox' ).checked ).toBeFalsy();
    } );

    test( 'Checkbox selected and css class doneToDo included', () => {
        const toDoItemCompleted = { ...toDoItem, [ 'done' ]: true };
        render( 
            <ToDoItem 
                { ...toDoItemCompleted }
                handleRemoveToDo={ handleRemoveToDoMock }
                handleToggleToDo={ handleToggleToDoMock }
            /> 
        );

        expect( screen.getByTestId( 'title-to-do' ).className ).toBe( 'doneToDo' );
        expect( screen.getByRole( 'checkbox' ).checked ).toBeTruthy();
    } );

    test( 'ToDoItem should call handleRemoveToDo function', () => {
        render( 
            <ToDoItem 
                { ...toDoItem }
                handleRemoveToDo={ handleRemoveToDoMock }
                handleToggleToDo={ handleToggleToDoMock }
            /> 
        );
        fireEvent.click( screen.getByRole( 'button' ) );

        expect( handleRemoveToDoMock ).toBeCalled();
        expect( handleRemoveToDoMock ).toBeCalledTimes( 1 );
        expect( handleRemoveToDoMock ).toBeCalledWith( toDoItem.id );
    } );

    test( 'ToDoItem should call handleToggleToDo function', () => {
        render( 
            <ToDoItem 
                { ...toDoItem }
                handleRemoveToDo={ handleRemoveToDoMock }
                handleToggleToDo={ handleToggleToDoMock }
            /> 
        );
        fireEvent.click( screen.getByRole( 'checkbox' ) );

        expect( handleToggleToDoMock ).toHaveBeenCalled();
        expect( handleToggleToDoMock ).toBeCalledTimes( 1 );
        expect( handleToggleToDoMock ).toBeCalledWith( toDoItem.id );
    } );
} );