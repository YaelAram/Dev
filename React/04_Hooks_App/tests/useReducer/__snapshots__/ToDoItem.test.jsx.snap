// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoItem component Match snapshot 1`] = `
<div>
  <li>
    <input
      type="checkbox"
    />
    <span
      class=""
      data-testid="title-to-do"
    >
      Hello World
    </span>
    <input
      type="button"
      value="Remove"
    />
  </li>
</div>
`;
